src/rust/src/incremental_merkle_tree.rs

```diff
@@ -1,120 +1,276 @@
     })?;
-    Vector::write_sized(
-        &mut writer,
-        tree.witnessed_indices().iter(),
-        |mut w, (pos, i)| {
-            write_position(&mut w, *pos)?;
-            write_usize_leu64(&mut w, *i)
-        },
-    )?;
-    Vector::write(&mut writer, tree.checkpoints(), |w, c| {
-        write_checkpoint_v2(w, c)
+    // mark positions forgotten due to notes at those positions having been
+    // spent since the position at which this checkpoint was created
+    Vector::write_sized(&mut writer, checkpoint.forgotten().iter(), |w, pos| {
+        write_position(w, *pos)
     })?;
-    write_usize_leu64(&mut writer, tree.max_checkpoints())?;
 
     Ok(())
 }
 
+/// Reads a [`BridgeTree`] value from its serialized form.
+///
+/// [`BridgeTree`] values are expected to have been serialized with a leading version byte. Parsing
+/// behavior varies slightly based upon the serialization version.
+///
+/// SER_V1 checkpoint serialization encoded checkpoint data from the `Checkpoint` type as defined
+/// in `incrementalmerkletree` version `0.3.0-beta-2`. This version was only used in testnet
+/// wallets prior to NU5 launch. Reading `SER_V1` checkpoint data is not supported.
+///
+/// Checkpoint identifiers are `u32` values which for `SER_V3` serialization correspond to block
+/// heights; checkpoint identifiers were not present in `SER_V2` serialization, so when reading
+/// such data the returned identifiers will *not* correspond to block heights. As such, checkpoint
+/// ids should always be treated as opaque, totally ordered identifiers without additional
+/// semantics.
 #[allow(clippy::redundant_closure)]
 pub fn read_tree<H: Hashable + HashSer + Ord + Clone, R: Read>(
     mut reader: R,
-) -> io::Result<BridgeTree<H, 32>> {
+) -> io::Result<BridgeTree<H, u32, MERKLE_DEPTH>> {
     let ser_version = reader.read_u8()?;
-    let prior_bridges = Vector::read(&mut reader, |r| read_bridge_v1(r))?;
-    let current_bridge = Optional::read(&mut reader, |r| read_bridge_v1(r))?;
+    let prior_bridges = Vector::read(&mut reader, |r| read_bridge(r))?;
+    let current_bridge = Optional::read(&mut reader, |r| read_bridge(r))?;
     let saved: BTreeMap<Position, usize> = Vector::read_collected(&mut reader, |mut r| {
         Ok((read_position(&mut r)?, read_leu64_usize(&mut r)?))
     })?;
 
     let checkpoints = match ser_version {
         SER_V1 => {
-            let mut save_positions = saved.iter().map(|(p, i)| (*i, *p)).collect();
-            Vector::read_collected_mut(&mut reader, |r| read_checkpoint_v1(r, &mut save_positions))
+            // SER_V1 checkpoint serialization encoded checkpoint data from the `Checkpoint` type
+            // as defined in `incrementalmerkletree` version `0.3.0-beta-2`. This version was only
+            // used in testnet wallets prior to NU5 launch. As such, we no longer support reading
+            // v1 checkpoint data.
+            Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
```

Comment:
```suggestion
                io::ErrorKind::InvalidData,
```
https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.InvalidData says:
> Unlike [`InvalidInput`](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.InvalidInput), this typically means that the operation parameters were valid, however the error was caused by malformed input data.
which is correct here.